src/Core/src/Signer/SignerV4.php

```diff
@@ -98,19 +110,19 @@ private function handleSignature(Request $request, Credentials $credentials, \Da
         $credentialString = \implode('/', $credentialScope);
         $signingKey = $this->buildSigningKey($credentials, $credentialScope);
 
-        // signature is passed by reference to convertBodyToStream
-        $signature = '';
         if ($isPresign) {
             // Should be called before `buildBodyDigest` because this method may alter the body
             $this->convertBodyToQuery($request);
         } else {
+            // $signature does not exists but passed by reference then computed buildSignature
+            $signature = '';
             $this->convertBodyToStream($request, $now, $credentialString, $signingKey, $signature);
```

Comment:
Why don’t we return the signature instead of passing it as reference?

Reply:
this method need the signature generated by the BuildSignature method Bellow.
And the BuildSignature method need headers generated by this method.
Their is a circular dependency between both method

Reply:
but then, does it work properly at all ?

Reply:
yes it works. Thanks to the reference.

here is the simplified workflow
```php
function convertBodyToStream() {
  $request->setHeader('xxx'); // change something in the header

  // request getBody will be called later when httpClient will sent the request
  $request->setBody(function() use ($signature) {
   //do someting with the signature
  });
}
function buildSignature()
{
  // need the headers to build the signature
  $signature = xx($request->getHeaders(), $args)
}
```